runtime/libc/wasi_sdk_backing.c

```diff
@@ -27,188 +28,95 @@
+ */
+void runtime_cleanup() {
+    return;
+}
 
-// some definitions we need
-struct wasi_iovec {
-    i32 base_offset;
-    i32 len;
-};
+/**
+ * @brief Copies arguments provided to the runtime process into global data structures accessible
+ * to the WASI argument syscalls
+ * 
+ * @param argc 
+ * @param argv 
+ */
+void runtime_args_init(int argc, char* argv[]) {
+    /* Set argc and argv to globals, these are later used by the WASI syscalls */
+    runtime_argc = argc;
+    runtime_argv_buffer_offsets = calloc(argc, sizeof(wasi_size_t));
+    if (runtime_argv_buffer_offsets == NULL) {
+        fprintf(stderr, "Error allocating runtime_argv_buffer_offsets: %s", strerror(errno));
+        exit(EXIT_FAILURE);
+    }
+    atexit(runtime_argv_buffer_offsets_free);
 
-struct wasi_fdstat {
-    u8  fs_filetype;
-    u16 fs_flags;
-    u64 fs_rights_base;
-    u64 fs_rights_inheriting;
-};
+    int i;
 
-struct wasi_filestat {
-    u64 st_dev;
-    u64 st_ino;
-    u8 st_filetype;
-    u32 st_nlink;
-    u64 st_size;
-    u64 st_atim;
-    u64 st_mtim;
-    u64 st_ctim;
-};
+    /* Calculate vector of argument offsets and argument buffer length */
+    for (i = 0; i < argc; i++) {
+        runtime_argv_buffer_offsets[i] = runtime_argv_buffer_len;
+        runtime_argv_buffer_len += (strlen(argv[i]) + 1);
+    }
 
-enum wasi_filetype {
-    WASI_FILETYPE_UNKNOWN           = 0,
-    WASI_FILETYPE_BLOCK_DEVICE      = 1,
-    WASI_FILETYPE_CHARACTER_DEVICE  = 2,
-    WASI_FILETYPE_DIRECTORY         = 3,
-    WASI_FILETYPE_REGULAR_FILE      = 4,
-    WASI_FILETYPE_SOCKET_DGRAM      = 5,
-    WASI_FILETYPE_SOCKET_STREAM     = 6,
-    WASI_FILETYPE_SYMBOLIC_LINK     = 7,
-};
+    /* Allocate argument buffer */
+    runtime_argv_buffer = malloc(runtime_argv_buffer_len);
+    if (runtime_argv_buffer == NULL) {
+        fprintf(stderr, "Error allocating runtime_argv_buffer: %s", strerror(errno));
+        exit(EXIT_FAILURE);
+    }
+    atexit(runtime_argv_buffer_free);
+        
+    /* Copy the arguments consecutively into the buffer based on the offsets */
+    for (i = 0; i < argc - 1; i++) {
+        strncpy(&runtime_argv_buffer[runtime_argv_buffer_offsets[i]], argv[i], runtime_argv_buffer_offsets[i + 1] - runtime_argv_buffer_offsets[i]);
+    }
+    if (argc > 0) {
+        strncpy(&runtime_argv_buffer[runtime_argv_buffer_offsets[i]], argv[i], runtime_argv_buffer_len - runtime_argv_buffer_offsets[i]);
+    }
+}
 
-enum wasi_oflags {
-    WASI_O_CREAT        = 0x0001,
-    WASI_O_DIRECTORY    = 0x0002,
-    WASI_O_EXCL         = 0x0004,
-    WASI_O_TRUNC        = 0x0008,
-};
+int main(int argc, char* argv[]) {
+    runtime_init();
+    runtime_args_init(argc, argv);
 
-enum wasi_fdflags {
-    WASI_FDFLAG_APPEND      = 0x0001,
-    WASI_FDFLAG_DSYNC       = 0x0002,
-    WASI_FDFLAG_NONBLOCK    = 0x0004,
-    WASI_FDFLAG_RSYNC       = 0x0008,
-    WASI_FDFLAG_SYNC        = 0x0010,
-};
+    atexit(runtime_cleanup);
```

Comment:
I find the multiple `atexit` calls somewhat confusing

Reply:
It's fine as is, but it also seems we could put the buffer freeing code in "runtime_cleanup"

Not a blocker for PR

Reply:
For error handling associated with globals, I prefer the multiple atexit handlers because they are effectively pushed to a stack. That means if there is some sort of fault within `runtime_init`, only the dynamic allocations that succeeded end up getting cleaned up. It believe this also mirrors the way GOTOs and labels are used for local error handling in the Composite style guide.

I renamed `runtime_cleanup` to `runtime_on_module_exit` to clarify that this cb is mostly intended for debugging / logging, not cleanup of runtime globals. I also moved `switch_into_runtime` into the `proc_exit` syscall to reduce the atexit calls a bit.